special-case #[derive(Copy, Clone)] with a shallow clone

durka opened the pull request:
If a type is Copy then its Clone implementation can be a no-op. Currently `#[derive(Clone)]` generates a deep clone anyway. This can lead to lots of code bloat.

This PR detects the case where Copy and Clone are both being derived (the general case of "is this type Copy" can't be determined by a syntax extension) and generates the shallow Clone impl. Right now this can only be done if there are no type parameters (see https://github.com/rust-lang/rust/issues/31085#issuecomment-178988663), but this restriction can be removed after specialization.

Fixes #31085.

rust-highfive commented:
r? @brson

(rust_highfive has picked a reviewer for you, use r? to override)

nagisa commented on src/libsyntax_ext/deriving/clone.rs:
```diff
@@ -24,13 +25,54 @@ pub fn expand_deriving_clone(cx: &mut ExtCtxt,
+                ast::ItemStruct(_, ast::Generics { ref ty_params, .. })
+                    | ast::ItemEnum(_, ast::Generics { ref ty_params, .. })
+                    if ty_params.is_empty()
+                        && (|| item.attrs.iter().any(|a| a.name() == "derive_Copy"))() => {
```
syntax::attr::contains_name(item.attrs, "derive_Copy")

nagisa commented on src/libsyntax_ext/deriving/clone.rs:
```diff
@@ -24,13 +25,54 @@ pub fn expand_deriving_clone(cx: &mut ExtCtxt,
+            }
+        }
+
+        _ => cx.span_bug(span, "#[derive(Clone)] on trait item or impl item")
```
So this branch is not trigerred for non-trait/impl items? Consider: functions, `type` aliases, `mod` items, etc?

durka replied:
AFAIK those should never reach this path because they error out here: https://github.com/rust-lang/rust/blob/master/src/libsyntax_ext/deriving/mod.rs#L122

durka replied:
Oh, I misinterpreted what you meant. There are only three choices for Annotatable: http://manishearth.github.io/rust-internals-docs/syntax/ext/base/enum.Annotatable.html

durka replied:
I'm wrong again because it could be a different kind of Item. So it should just ICE with "... on non-struct/enum item"

nagisa replied:
Right, I’m surprised this branch is necessary, when any sort of item “files” just fine behind the `Item` variant.

durka replied:
No, I'm going back to my second comment. There's a nested match. This unreachable catchall arm is in the outer one, which differentiates between Item vs TraitItem/ImplItem. The inner one does the check to see which kind of Clone we can derive.

nagisa replied:
> I'm wrong again because it could be a different kind of Item. So it should just ICE with "... on non-struct/enum item"

I think you should just have something like

```
match *item {
    Annotatable::Item(ast::Item { node: ast::ItemStruct { ... } }) |
    Annotatable::Item(ast::Item { node: ast::ItemEnum { ... } })
    if /* the other conditions */ => {
         // this is copy branch
    }
    _ => /* the clone branch */
}
```

durka replied:
It doesn't work because it's `enum Annotatable { Item(P<Item>), ... }` and you can't match through a `P`. At least, I can't. I tried, with `#![feature(box_patterns)]` but the `ptr` field is private so I can't get to the inner `Box`.

bluss commented:
According to https://github.com/rust-lang/rust/issues/31085#issuecomment-173743819 this kind of improvement can improve compile times (that's a case with a lot of structs).

According to me @ https://github.com/rust-lang/rust/issues/31085#issuecomment-173684981 using the Copy-based clone can improve codegen for loops calling .clone().

bors commented:
:umbrella: The latest upstream changes (presumably #31400) made this pull request unmergeable. Please resolve the merge conflicts.

durka commented:
It occurs to me that the shallow Clone can also be done in this situation:

```rust
#[derive(Copy, Clone)]
struct S<T: Copy>(T);
```

i.e. when there _are_ generic parameters _but_ they are all already constrained to be Copy. Is it worth it to add this check as well?

durka commented:
Rebased.

nagisa commented:
This PR LGTM, but I’d like an opinion from somebody more familiar with the code as well.

brson commented on src/test/run-make/pretty-expanded-hygiene/input.rs:
```diff
@@ -16,6 +16,12 @@ macro_rules! foo {
+
+#[derive(Copy, Clone)]
+struct B<T> { a: T }
+
```
Can you put this in its own rmake test case instead of piggy-backing on this one?

durka replied:
I'll split the tests and revert my changes to the Makefile for this one.

brson commented:
Do we have any concrete numbers that this gives to either compile time or runtime? @retep998 said he worked around this on winapi because of bad compile times. Perhaps we can change winapi to use deriving again and see if this improves perf there.

durka commented:
Travis failed because I didn't run the full test suite locally. Will fix later tonight.

durka commented:
I split the tests and tried to fix the fallout in the rpass-full tests. I'm getting linker errors locally which I think will be solved by `make clean`, we'll see.

durka commented:
@brson @bluss here are the runtime benchmarks: [gist](https://gist.github.com/durka/206b1a04ff57ea4c95fd)

The performance of `clone_from_slice` on `#[derive(CopyClone)] struct S(u8, u32);` is doubled.

durka commented:
@brson @retep998 and compile time: [gist](https://gist.github.com/durka/5b061865e9b2a968857f)

A `#[derive(Copy, Clone)]` struct with 1000 members compiles 13 times faster.

durka commented:
As a bonus, this extends the reach of `#[derive(Clone)]` -- you can now `#[derive(Copy, Clone)]` on structs containing types which are Copy but not Clone, such as `[u8; 100]`. [gist](https://gist.github.com/durka/9db77fd583be6dcf8406)

bluss commented:
Compile time win for structs and enums might be the best thing with this. A bit unnerving that it enables deriving Clone on more things, though. That makes it hard to go back on this optimization.

durka commented:
@bluss I don't see why we'd want to go back on it. Plus, more motivation to fix the issue where types can be magically Copy but not Clone :)

brson commented:
Wow, great numbers.

I'm also uneasy about the semantic change this brings, making derive Copy/Clone work on types containing those that aren't clone. It exposes the optimization - semantically we may actually want derive Clone to be defined *as if* it was implemented the naive way.

I assume this doesn't magically make just `derive(Clone)` work for types containing types that don't implement `Clone`, but only `derive(Copy, Clone)`, and that's super weird.

durka commented:
> I assume this doesn't magically make just `derive(Clone)` work for types containing types that don't implement `Clone`, but only `derive(Copy, Clone)`

Correct. The derives just do their little syntactical shufflings and leave it to typeck to enforce the constraints later.

> and that's super weird.

I agree, *but* it's only observable if you have a type that's `Copy` and not `Clone`, which isn't supposed to be possible. So it only comes up with corner cases like arrays over size 32 or tuples over size 12.

durka commented:
We could revert the semantic change by generating code like this:

```rust
impl Clone for Foo {
    fn clone(&self) -> Foo {
        if false { /* naive implementation */ }
        else { *self }
    }
}
```

but we'd of course lose the compile-time wins.

sfackler commented:
I think we optimize `if false ...` in the frontend so that branch would at least not make it to LLVM.

Can't we also do something like

```rust
struct Foo {
    bar :Bar,
}

// generated
impl Clone for Foo {
    fn clone(&self) -> Foo {
        fn is_clone<C: Clone>() {}
        is_clone::<Bar>();
        *self
    }
}
```

which should avoid the compile time sadness.

sfackler commented:
That does still leave the behavior change that `bar`'s `Clone` impl isn't invoked obviously.

durka commented:
@sfackler I'm trying this. I still think it's insignificant since the existence of `T: Copy + !Clone` is a bug.

bluss commented:
Every existence of a Copy + !Clone value can be used to provoke an ICE, so it's clearly a bug we need to solve eventually. I think the simple `*self` version is ok for this reason.

durka commented:
I implemented the if-false version. It does revert the expanded reach of `#[derive(Copy, Clone)]`.

Runtime benchmark is unaffected. However the compile time difference disappears.

bors commented:
:umbrella: The latest upstream changes (presumably #31487) made this pull request unmergeable. Please resolve the merge conflicts.

durka commented:
Rebased and fixed for enumpocalypse. if-false sub-branch moved to [here](https://github.com/durka/rust/compare/clone-copy...durka:clone-copy-if-false).

brson commented:
> I agree, but it's only observable if you have a type that's Copy and not Clone, which isn't supposed to be possible. So it only comes up with corner cases like arrays over size 32 or tuples over size 12.

These types are possible and they exist. I don't think we should introduce these special semantics for `#[derive(Copy, Clone)]` on unclonable types. It would create a weird special case in the language spec. Can this optimization be done without it?

durka commented:
They exist because of a bug, so I guess I see the weird special case going the other way. But anyway, as above it looks like the runtime optimization can be done without changing semantics, but not the compile time, unless someone has a cleverer idea than if-false for me to try.

durka commented:
It sounds like specialization could obsolete this, eventually (i.e. `impl<T: Copy> Clone for T { fn clone(&self) -> Self { *self } }`), but not with the "simple subset" rule that is currently proposed, only if that is extended.

bluss commented:
@brson 

> These types are possible and they exist.

On the other hand their existence is a bug, and every instance of them can be used to trigger the ICE in #25733 (not just the arrays, but all such Copy + !Clone types.

briansmith commented:
> If a type is Copy then its Clone implementation can be a no-op. Currently #[derive(Clone)] generates a deep clone anyway.

First, I am not objecting to this change. I'm just pointing out it is a special-case of a larger issue:

Why didn't the optimizer optimize the generated code to `memcpy`? There are lots of coding patterns, including the coding pattern that is generated for `clone()`, that can be optimized to memcpy and it seems like the Rust compiler (through LLVM) already optimizes some of them that way. For example, I've written some loops that never call memcpy but which get compiled into memcpy. It seems to me that it is likely that there is something wrong with the optimizer if it osn't doing this for `clone` implementations, and that means it probably is missing a lot of other cases too.

briansmith commented:
In particular, consider the case of the functional record update pattern, where an object is copied except that one field is changed. It is likely that many instances of that pattern need to be optimized to `memcpy` + a single field write in order to get reasonable performance.

Thus, it definitely seems worthwhile to investigate a way to get the optimizer to do this optimization more generally without any intrinsic knowledge of `Copy` or `Clone`.

durka commented:
Okay, this is stalled. So let's try to take it in a productive direction.

**Situation**:

1. There are some types that are `Copy` but not `Clone`. This is in violation of the declaration of `Copy` in libcore, but it is true nonetheless.
2. At some point, this situation will need to be fixed. This is because it causes ICEs as cited above.
3. I see only two possible ways to fix this:
    1. These types will stop being `Copy`.
    2. These types will become `Clone` as well as `Copy`.
4. 3(i) is a breaking change, and we are generally not about those, so 3(ii) seems more likely. When that happens, *however* it happens, `#[derive(Clone, Copy)]` will become available to structs/enums containing those types -- the very same structs/enums that the current PR would (accidentally, hackishly) affect. It *could* happen by specialization, if it's extended such that we can write `impl<T: Copy> Clone for T`. In that case, the entire optimization done by this PR is subsumed. But I don't know how to comment on the likeliness of specialization being extended.

**Questions**:

- In what other way(s) can the `Copy + !Clone` situation can be fixed if specialization doesn't cover it? Are there other ideas?
- Should we close this PR as killed by technicality, or leave it open for more back-and-forth?

glaebhoerl commented:
(I am personally baffled by the logic of rejecting a performance optimization just because it accidentally includes an incomplete bugfix as well...)

petrochenkov commented:
>(I am personally baffled by the logic of rejecting a performance optimization just because it accidentally includes an incomplete bugfix as well...)

+

bors commented:
:umbrella: The latest upstream changes (presumably #31822) made this pull request unmergeable. Please resolve the merge conflicts.

ubsan commented:
The existence of `T: Copy + !Clone` is a bug. The fact that this fixes one aspect of that bug, and not most of it, shouldn't be super important, imho. Currently, there's a major pain point where people have to manually `impl Clone` for types with large `Copy` arrays inside them, which seems silly when 1) we can fix it now, and 2) this is the intended semantics of `Clone` for `T: Copy`.

bors commented:
:umbrella: The latest upstream changes (presumably #32251) made this pull request unmergeable. Please resolve the merge conflicts.

brson commented:
To those that see this as a bug fix that should happen regardless, the way I see it is that this optimization is a layering violation - the bug of `Clone` not being implemented correctly should not be papered over by a macro in hopes that one day a proper fix will happen. If that fix never happens then the macro system will forever have a special case that disagrees with what's possible in the language and libraries.

Regardless I've escalated for further discussion.

retep998 commented:
If that fix never happens then I will be quite disappointed in Rust. Having built in types violating the type system just isn't right.

SimonSapin commented:
Could the optimizer be taught to better recognize code that looks like derived `Clone` (structs literals and enum contructors where all fields come from something with the same layout) and replace it with `memcpy` when that doesn’t change semantics?

brson commented:
After talking with the libs team we

pnkfelix commented:
triage: I-nominated

(I want the lang team to discuss this)

pnkfelix commented:
(actually I think the compiler team may also be able to have a useful discussion about the part about whether we can improve the optimizer to recognize the code patterns here, so I'm going to add T-compiler to the list...)

brson commented:
After discussion with the libs team we'd like to add a hack to verify that all fields are `Clone`. This is similar to what `derive(Eq)` does to validate that fields implement `Eq`. The way to do this is to write something like the following in the emitted methods:

```
    fn assert_clone<T: Clone>(_t: &T) {}
    assert_clone(&self.field1);
```

This will prevent us from emitting implementations where the fields aren't `Clone`.

pnkfelix commented:
@brson I assume you want that `assert_clone` thing emitted for all `#[derive(Copy)]`, not just the ones that do `#[derive(Copy, Clone)]` ?

(in other words, can't that hack be considered independently from this PR?)

alexcrichton commented:
Yes to clarify from the discussion we had, we're wary about basically enabling more features that weren't possible today (for whatever reason), so we want `#[derive(Copy, Clone)]` to have the same limitations tomorrow as it does today. That essentially means that we'll just require that all fields implement the `Clone` trait themselves.

In light of that I think there's no real change here other than exactly what's accepted today is still accepted (same with rejections) and the generated code should be faster to codegen and faster in debug mode.

durka commented:
Do you like the [`if false` hack](https://github.com/rust-lang/rust/pull/31414#issuecomment-180891553) as [implemented already](https://github.com/durka/rust/commit/7aea0ef7d23b590e25de0e882cb4f98cdc2ca73d#diff-87c5c61e46351211f06181db35ff6cd4L20) or should I try the `assert_clone` version as well?

durka commented:
Perhaps `assert_clone` could be faster to compile though? I guess there's only one way to find out.

alexcrichton commented:
In terms of implementation that seems fine to me, yeah. Could you confirm that @sfackler's [suspicions](https://github.com/rust-lang/rust/pull/31414#issuecomment-181042481) about codegen are indeed correct, though?

If we end up just generating code for both branches then we likely want a solution like `assert_clone`

durka commented:
@alexcrichton when I tried it before the if-false cause compile time to go back up to what it was before -- i.e. `if false` is not optimized before codegen.

alexcrichton commented:
Sounds like we should explore the `assert_clone`-style solution instead then?

durka commented:
Yeah I will do some experiments and report back.

On Thu, Mar 24, 2016 at 2:45 PM, Alex Crichton <notifications@github.com>
wrote:

> Sounds like we should explore the assert_clone-style solution instead
> then?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rust/pull/31414#issuecomment-200965696>
>

sfackler commented:
Is this ready to go?

durka commented:
I'm gathering benchmarks on the various ways to make it not expand the
reach of deriving. They're almost ready, sorry for taking so long.
On Apr 9, 2016 22:07, "Steven Fackler" <notifications@github.com> wrote:

> Is this ready to go?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rust/pull/31414#issuecomment-207897466>
>

durka commented:
OK, I finally got back to this.

I've implemented both the `if false` and `assert_clone` solutions. You can see the differences in output from `#[derive(Copy, Clone)]`: [`no hack`](), [`if false`](), [`assert_clone`]().

I have conducted some [microbenchmarking](https://gist.github.com/durka/7c8867a554e51dff9c4f995da3bad918) using (a) a struct with 1000 fields and (b) 1000 structs with one field (which would you rather compile?). These tests were conducted with the stage1 compiler, optimizations on, on x86_64-unknown-linux-gnu, single core processor.

| | master | no hack | `if false` | `assert_clone ` |
| --- | --- | --- | --- | --- |
| *compile time (1000 fields)* | 3.771s | 0.737s | 2.138s | 1.898s |
| *runtime (1000 fields)* | 6.577s | 13.564s | 14.213s | 14.029s |
| *binary size (1000 fields)* | 925.78K | 925.76K | 925.76K | 925.76K |
| *compile time (1000 structs)* | 4m52.975s | 5m20.204s | 5m41.629s | 5m2.783s |
| *runtime (1000 structs)* | 0.900s | 0.898 | 0.922s | 0.966s |
| *binary size (1000 structs)* | 957.78K | 957.76K | 957.76K | 957.76K |

I can't really explain these results. Maybe my benchmark is not measuring what I think it is? Why are the runtimes always slower than master? Perhaps something happening with vectorization.

bluss commented:
The horse_sized_ducks.rs and Clone uses the fact that it knows all fields are zero.

You can do this instead: `let duck = test::black_box(Duck::default());` and it can't do that anymore (this does indeed flip the difference in runtime to favour the new code instead.).

durka commented:
I knew I could count on @bluss to explain what optimizations were going on. I've updated the table of results.

So it seems that runtime is a wash. `assert_clone` seems a bit better on compile time in the many-fields case, but there's no real difference, or possibly the opposite, in the many-structs case. I could also try putting the `assert_clone` method on the `Clone` trait itself, like `Eq::assert_receiver_is_total_eq`.

bluss commented:
assert_clone seems a lot more sane than the if false approach IMO. It doesn't have all the wins of the naked no hack approach, but it's better than the current derive code in every way anyway? I think it looks good if we can tolerate new kludges like assert_clone.

bluss commented:
Not declaring assert_clone over and over again should save compilation time, right?

durka commented:
Yeah, maybe. I'll check it out. Also has anyone tried for-real specializing `Clone` for `T: Copy`? Though I know it's not supposed to work with the current `#![feature(specialization)]`.

alexcrichton commented:
Thanks for gathering the data @durka! I agree with @bluss that the best approach here seems to be `assert_clone`. I'd be surprised if there were wins from consolidating calls to `assert_clone`, but hey you never know!

durka commented:
Declaring the method once did seem to shave a few seconds off the compile time. I updated the table of results and the gists again.

durka commented:
I've swapped the branches here so that `assert_clone` is the PR branch, the old "pure" version is at durka/clone-copy-no-hack.

durka commented:
It looks like I messed up some of the error spans.

durka commented:
Gah, now the syntax extension tests are broken. I might not get to fix that
until Monday. And I guess this is now a libsyntax-breaking-change.
On Apr 12, 2016 01:20, "Alex Crichton" <notifications@github.com> wrote:

> Thanks for gathering the data @durka <https://github.com/durka>! I agree
> with @bluss <https://github.com/bluss> that the best approach here seems
> to be assert_clone. I'd be surprised if there were wins from
> consolidating calls to assert_clone, but hey you never know!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rust/pull/31414#issuecomment-208711681>
>

durka commented:
Woohoo, tests passed!

Manishearth commented:
cc https://github.com/rust-lang/rust/issues/31645

Manishearth commented:
Please wait a few more days before landing. This should be landed simultaneously with https://github.com/rust-lang/rust/pull/33041.

bluss commented on src/libcore/clone.rs:
```diff
@@ -75,6 +75,15 @@ pub trait Clone : Sized {
+#[doc(hidden)]
+#[inline(always)]
+#[stable(feature = "rust1", since = "1.0.0")]
+pub fn assert_receiver_is_clone<T: Clone + ?Sized>(_: &T) {}
```
should this be stable or unstable? It needs a new feature name and updated version.

durka replied:
Yeah, it should be unstable.

Manishearth commented:
Is this ready to land?

r? @nagisa 

(don't bors:r+ this, just mention r=me somewhere)

durka commented:
I'm making `core::clone::assert_receiver_is_clone` unstable.

durka commented:
Also, I noticed that `#[derive(Eq)]` does something similar, so maybe there is a way to use the generic stuff already in place without a breaking change.

Manishearth commented:
@eddyb @brson @alexcrichton Could this be reviewed? It's blocking a bunch of other PRs due to breaking-batch.

eddyb commented on src/libsyntax_ext/deriving/clone.rs:
```diff
@@ -11,40 +11,87 @@
+                ItemKind::Struct(_, Generics { ref ty_params, .. }) |
+                ItemKind::Enum(_, Generics { ref ty_params, .. })
+                    if ty_params.is_empty()
+                        && attr::contains_name(&annitem.attrs, "derive_Copy") => {
```
Doesn't this depend on the ordering of `Clone` and `Copy` in the `#[derive]` attribute?

durka replied:
It shouldn't, but I should probably add a test. It's checking for `#[derive(Copy)]` while processing `#[derive(Clone)]`, they are already separate attributes by this point AIUI.

eddyb commented:
Can this be implemented with specialization *one a case-by-case basis*? i.e. not for all types, but just for every `impl` produced by `#[derive(Clone)]`.

durka commented:
Probably. Do we want to introduce that much specialization into everyone's generated code?

eddyb commented:
@durka Good question, it would likely make winapi worse, not better. Maybe replace `#[derive(Copy, Clone)]` with `#[derive_CopyClone]` or something like that?

durka commented:
@Manishearth sorry about my slowness. I still need to fix the spans so I
can make core::assert_receiver_is_clone unstable. And I was hoping if I
stared at the deriving code long enough I could remove the breakage, but I
don't see how. So punt on this if you want, otherwise I'll really try to
get it ready to land this weekend.
On Apr 22, 2016 17:36, "Eduard-Mihai Burtescu" <notifications@github.com>
wrote:

> @durka <https://github.com/durka> Good question, it would likely make
> winapi worse, not better. Maybe replace #[derive(Copy, Clone)] with
> #[derive_CopyClone] or something like that?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rust/pull/31414#issuecomment-213595559>
>

Manishearth commented:
Nah, it's okay, we can wait a few days :smile: 

I wasn't sure what the status of this PR was.

retep998 commented:
@eddyb I don't use `#[derive(Clone)]` at all in winapi.

durka commented:
@retep998 but you could in a parallel universe where #[derive(Clone)]
didn't slow things down and winapi didn't care about 1.x backcompat.
On Apr 22, 2016 21:51, "Peter Atashian" <notifications@github.com> wrote:

> @eddyb <https://github.com/eddyb> I don't use #[derive(Clone)] at all in
> winapi.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rust/pull/31414#issuecomment-213647649>
>

retep998 commented:
@durka But I can't due some `Copy` types not being `Clone` such as function pointers and large arrays.

bluss commented:
How lovely, winapi needs the bugs that this PR takes care to not expose. :smile: Can't please them all

durka commented:
The stage1-rmake test passes now. Still waiting on the stage2-rpass-full one. And I guess travis will try them all.

durka commented:
@eddyb and I are now certain we can do this without the breaking change. un-cc #31645 and my apologies for delaying the batch :)

durka commented:
Breaking change removed. Performance gains retained.

alexcrichton commented on src/test/run-pass/issue-3121.rs:
```diff
@@ -8,7 +8,7 @@
 // except according to those terms.
 
-// pretty-expanded FIXME #23616
+// no-pretty-expanded FIXME #23616 assert_receiver_is_clone
```
Instead of `no-pretty-expanded` I think you can just remove these directives altogether

durka replied:
Oh is it opt-in? Ok. I thought it was opt-out. I also figured if I left the mention of the method name here, then someone in the future stabilizing or removing this hack might grep for it and realize they could re-enable the test.

alexcrichton replied:
Nah yeah it's fine to just remove everywhere, right now that phase of testing is opt-in

durka replied:
Fair enough. Does that mean when I write a new run-pass test I should try
to add "// pretty-expanded"? Do we have a policy on that?

On Mon, Apr 25, 2016 at 2:24 PM, Alex Crichton <notifications@github.com>
wrote:

> In src/test/run-pass/issue-3121.rs
> <https://github.com/rust-lang/rust/pull/31414#discussion_r60962453>:
>
> > @@ -8,7 +8,7 @@
> >  // option. This file may not be copied, modified, or distributed
> >  // except according to those terms.
> >
> > -// pretty-expanded FIXME #23616
> > +// no-pretty-expanded FIXME #23616 assert_receiver_is_clone
>
> Nah yeah it's fine to just remove everywhere, right now that phase of
> testing is opt-in
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rust/pull/31414/files/64d22d453d5b71b33b149dac3c28b2d626446d9f#r60962453>
>

alexcrichton replied:
Nah it's fine to just remove these comments

durka replied:
Fixed.

alexcrichton commented on src/test/run-make/expand-derive/input_pp.rs:
```diff
@@ -0,0 +1,51 @@
+    fn clone(&self) -> A {
+        match *self {
+            A { a: ref __self_0_0 } => {
+                ::clone::assert_receiver_is_clone(&(*__self_0_0));
```
I think we have `src/test/pretty` for this, right? Always just good to avoid run-make wherever possible :)

durka replied:
Hmm what does `src/test/pretty` actually do? This test asserts that the pretty expansion is exactly as intended but `src/test/pretty` doesn't seem to have expected output.

eddyb replied:
There's one `.pp` file in there which is the expected output for one test. If nobody has removed it yet for being painful to update, that is.

durka replied:
Oh, so there is. I can migrate this test to `src/test/pretty` then. But
there are more run-make tests of the same class (pretty-expanded,
pretty-expanded-hygiene). I guess they can be migrated in a separate PR.

On Mon, Apr 25, 2016 at 2:04 PM, Eduard-Mihai Burtescu <
notifications@github.com> wrote:

> In src/test/run-make/expand-derive/input_pp.rs
> <https://github.com/rust-lang/rust/pull/31414#discussion_r60958975>:
>
> > +
> > +// minimal junk
> > +#![feature(no_core)]
> > +#![no_core]
> > +
> > +struct A {
> > +    a: i32,
> > +}
> > +#[automatically_derived]
> > +#[allow(unused_qualifications)]
> > +impl ::clone::Clone for A {
> > +    #[inline]
> > +    fn clone(&self) -> A {
> > +        match *self {
> > +            A { a: ref __self_0_0 } => {
> > +                ::clone::assert_receiver_is_clone(&(*__self_0_0));
>
> There's one .pp file in there which is the expected output for one test.
> If nobody has removed it yet for being painful to update, that is.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rust/pull/31414/files/64d22d453d5b71b33b149dac3c28b2d626446d9f#r60958975>
>

alexcrichton replied:
Yeah I'd also be fine just leaving this test out, it'd end up being pretty brittle :(

durka replied:
Sure. `compile-fail/deriving-copyclone` is the important one for this PR.

durka replied:
I'm trying to think of a way to test the shallowness of a `Clone` impl without pp-exactness.

durka replied:
I think [this test](http://is.gd/CfUfEL) will do the job.

durka replied:
In other words this is a behavior change, but it is now OKed by accepted rust-lang/rfcs#1521.

alexcrichton replied:
Yeah that seems like a good test to me.

durka replied:
OK, I deleted the pretty-printing test and added the Innocent(Liar) test.

alexcrichton commented:
Thanks again for the PR @durka! Everything looks good to me, wanna squash and I'll r+?

durka commented:
Squashed and wrote a more extensive commit message.

On Tue, Apr 26, 2016 at 1:06 PM, Alex Crichton <notifications@github.com>
wrote:

> Thanks again for the PR @durka <https://github.com/durka>! Everything
> looks good to me, wanna squash and I'll r+?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rust/pull/31414#issuecomment-214813219>
>

alexcrichton commented:
@bors: r+ 9249e6a1e2ef57bb6e329e6477beed31647236b1

bors commented:
:hourglass: Testing commit 9249e6a with merge 8f55218...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/8936), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/8971), [auto-linux-32cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-32cross-opt/builds/61), [auto-linux-64-cargotest](http://buildbot.rust-lang.org/builders/auto-linux-64-cargotest/builds/253), [auto-linux-64-cross-armhf](http://buildbot.rust-lang.org/builders/auto-linux-64-cross-armhf/builds/61), [auto-linux-64-cross-armsf](http://buildbot.rust-lang.org/builders/auto-linux-64-cross-armsf/builds/64), [auto-linux-64-cross-freebsd](http://buildbot.rust-lang.org/builders/auto-linux-64-cross-freebsd/builds/64), [auto-linux-64-cross-netbsd](http://buildbot.rust-lang.org/builders/auto-linux-64-cross-netbsd/builds/64), [auto-linux-64-debug-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/2203), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/8910), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/8898), [auto-linux-64-opt-mir](http://buildbot.rust-lang.org/builders/auto-linux-64-opt-mir/builds/476), [auto-linux-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-linux-64-opt-rustbuild/builds/876), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/8893), [auto-linux-cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/2106), [auto-linux-musl-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/4051), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/8944), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/8955), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/8915), [auto-mac-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-mac-64-opt-rustbuild/builds/877), [auto-mac-cross-ios-opt](http://buildbot.rust-lang.org/builders/auto-mac-cross-ios-opt/builds/64), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/3975), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/3990), [auto-win-gnu-32-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt-rustbuild/builds/881), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/3986), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/3964), [auto-win-msvc-32-cross-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-cross-opt/builds/347), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/3182), [auto-win-msvc-64-cargotest](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-cargotest/builds/259), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/3976), [auto-win-msvc-64-opt-mir](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-mir/builds/479), [auto-win-msvc-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-rustbuild/builds/889)
